p2p/net/conn/dial.go

```diff
@@ -3,23 +3,32 @@ package conn
 import (
 	"fmt"
 	"math/rand"
-	"net"
 	"strings"
-	"syscall"
 
 	ma "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-multiaddr"
 	manet "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-multiaddr-net"
-	reuseport "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-reuseport"
 	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
 	lgbl "github.com/ipfs/go-ipfs/util/eventlog/loggables"
 
+	ci "github.com/ipfs/go-ipfs/p2p/crypto"
 	addrutil "github.com/ipfs/go-ipfs/p2p/net/swarm/addr"
+	transport "github.com/ipfs/go-ipfs/p2p/net/transport"
 	peer "github.com/ipfs/go-ipfs/p2p/peer"
 )
 
+type WrapFunc func(transport.Conn) transport.Conn
+
+func NewDialer(p peer.ID, pk ci.PrivKey, wrap WrapFunc) *Dialer {
```

Comment:
yeah `secio` should just be a _transport_ too, and when enabled it just wraps all the other conns (not specified in mutiadd). this is how it's working in js land and it's very good.